Adds map previews to the new dashboard

javierarce opened the pull request:
This branch adds map previews to the new dashboard using our Static Map API:

- the [_renderMapThumbnail](https://github.com/CartoDB/cartodb/pull/2078/files#diff-3820350fae17cd3cc619002993d7b83eR97) method requests a thumbnail using the visualization vizjson and updates the default image placeholder with the returned URL. 
- In case of error, [a random error image is shown instead](https://github.com/CartoDB/cartodb/pull/2078/files#diff-3820350fae17cd3cc619002993d7b83eR116).
- Adds [lib/assets/javascripts/cartodb/common/map_previews.js](https://github.com/CartoDB/cartodb/blob/1560-new-dashboard-medusa/lib/assets/javascripts/cartodb/common/map_previews.js) to render the previews cards from the [new_public_maps.html.erb](https://github.com/CartoDB/cartodb/blob/1560-new-dashboard-medusa/app/views/admin/pages/new_public_maps.html.erb#L17)

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1517/))

javierarce commented:
@xavijam, @viddo: can you please review this PR?

viddo commented on app/assets/stylesheets/new_public_dashboard/public-body.css.scss:
```diff
@@ -16,3 +16,10 @@
 }
+
+// Animation function
+@include keyframes(pulsate) {
```
I think it's better located in the same file as where it's used (i.e. map-card.css.scss)

viddo commented on lib/assets/javascripts/cartodb/new_dashboard/maps/maps_item.js:
```diff
@@ -93,6 +94,56 @@ module.exports = cdb.core.View.extend({
     this.addView(graph.render());
   },
 
+  _renderMapThumbnail: function() {
```
I have several concerns related to this method:

1. Since this method is almost identical to to the new `map_previews.js` and is indeed intended for the same conceptual piece, why not extract this logic to a common view that can be reused for the two contexts? 
 - the parts that differs (width, height, vizjson URL) can with benefits be provided from each context.
 - this way it also would use the newer browserify module to encapsulate modules rather than putting them on global namespace. I would suggest to create it as `new_common/views/map_card_preview_view`, possibly with a view model to go with it to isolate the state changes and handling
1. keeping variables referencing DOM elements (like `$header`) in the same scope as closures are keen to create memory leaks, I would suggest to refactor this code to avoid this
1. why are there no tests for new code?

let me know if you need some clarifications or other help

javierarce replied:
Thanks for the review!

I'll add tests and try to extract the common parts of the `renderMapThumbnail` and `loadMapPreview` methods, but what do you mean by "can with benefits be provided"?

viddo replied:
sorry for the ambiguous comment.. I was referring to extracting the common logic but keeping the variable parts in their right context, providing them as params to the common piece. something like:
```js
// somwhere in maps_item.js:
mapCardPreview = new MapCardPreview({
  width: 300,
  height: 170,
  // ..
});
```

viddo commented on lib/assets/javascripts/cartodb/new_dashboard/maps/maps_item.js:
```diff
@@ -93,6 +94,56 @@ module.exports = cdb.core.View.extend({
+
+    var host = this.router.currentUserUrl.host();
+
+    var vizjson = host + "/api/v2/viz/" + this.model.get("id") + "/viz.json";
```
The purpose of adding the URL models (located in `new_common/urls/`) was to have a single canonical place where to construct URLs. The `.host` is only public because it's required for the router to do client-side routing. Thus, I would propose to encapsulate (and test) this as a new URL model (since it's unrelated to the current user url), by the look of it we should probably add a ApiUrl model with semantics like:
```js
var vizjson = this.router.apiUrl().toVizjson(this.model);
```

javisantana replied:
In my opinion that url should be in the visualization model (those 3 steps are not needed, just add complexity)

javisantana replied:
in indeed, we have a bunch of urls in visualization model

viddo replied:
fair points, especially if we already have URLs created there :+1:

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1538/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1543/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1544/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1546/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1547/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1563/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1564/))

javierarce commented:
I've updated the code with your suggestions, @viddo and @javisantana. Could you please review it again?

### Last changes

- Added custom class to generate the previews ([mapcard_preview.js](https://github.com/CartoDB/cartodb/blob/1560-new-dashboard-medusa/lib/assets/javascripts/cartodb/new_dashboard/mapcard_preview.js))
- Added a new method to return the vizjson path in the [vis model](https://github.com/CartoDB/cartodb/blob/1560-new-dashboard-medusa/lib/assets/javascripts/cartodb/models/vis.js#L369).
- Removed duplication in the new public dashboard (file [entry.js](https://github.com/CartoDB/cartodb/blob/1560-new-dashboard-medusa/lib/assets/javascripts/cartodb/new_public_dashboard/entry.js))
- Added [specs](https://github.com/CartoDB/cartodb/blob/1560-new-dashboard-medusa/lib/assets/test/spec/cartodb/new_dashboard/mapcard_preview.spec.js).

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1565/))

viddo commented:
:ok_hand::ok_hand::ok_hand:

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1569/))

javisantana commented:
:es:

javierarce commented:
Thanks! :)

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/1570/))
